### ISSUE TITLE
scalarToColor algorithm fails

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#511]
**Modified by dietmarw on 29 Mar 2011 17:40 UTC**
Dear colleague,
the function call
```
Modelica.Mechanics.MultiBody.Visualizers.Colors.scalarToColor(scalarToColor(T=331,T_min=273,T_max=320,colorMap={{i,i,i} for i in 1:64})
```
fails im Dymola 7.4 FD01 and Modelica 3.2
which is caused by the following algorithm (in `Modelica.Mechanics.MultiBody.Visualizers.Colors.scalarToColor`):
 ```
color:=colorMap[integer((size(colorMap, 1) - 1)/(T_max - T_min)*
                            min((max(T,T_min) - T_min), T_max) + 1),:];
```
I propose the following algorithm instead (less elegant but robust):
```
color:=colorMap[min(integer(ceil(1/(T_max-T_min)*((size(colorMap, 1)-1)*
    max(T,T_min)+T_max-T_min*size(colorMap, 1)))),size(colorMap, 1)),:];
```
I attached a script with several test calls.
Best regards,
Andreas

----

**Modified by dietmarw on 29 Mar 2011 16:08 UTC**
Dear colleague,
the function call
```
Modelica.Mechanics.MultiBody.Visualizers.Colors.scalarToColor(scalarToColor(T=331,T_min=273,T_max=320,colorMap={{i,i,i} for i in 1:64})
```
fails im Dymola 7.4 FD01 and Modelica 3.2
which is caused by the following algorithm:
```
color:=colorMap[integer((size(colorMap, 1) - 1)/(T_max - T_min)*
                            min((max(T,T_min) - T_min), T_max) + 1),:];
```
I propose the following algorithm instead (less elegant but robust):
```
color:=colorMap[min(integer(ceil(1/(T_max-T_min)*((size(colorMap, 1)-1)*
    max(T,T_min)+T_max-T_min*size(colorMap, 1)))),size(colorMap, 1)),:];
```
I attached a script with several test calls.
Best regards,
Andreas

----

**Reported by andreas&#46;heckmann on 29 Mar 2011 15:00 UTC**
Dear colleague,
the function call

Modelica.Mechanics.MultiBody.Visualizers.Colors.scalarToColor(scalarToColor(T=331,T_min=273,T_max=320,colorMap={{i,i,i} for i in 1:64})

fails im Dymola 7.4 FD01 and Modelica 3.2
which is caused by the following algorithm:

 color :=colorMap[integer((size(colorMap, 1) - 1)/(T_max - T_min)*
                            min((max(T,T_min) - T_min), T_max) + 1), :];

I propose the following algorithm instead (less elegant but robust):

color:=colorMap[min(integer(ceil(1/(T_max-T_min)\*((size(colorMap, 1)-1)\*max(T,T_min)+T_max-T_min\*size(colorMap, 1)))),size(colorMap, 1)),:];

I attached a script with several test calls.
Best regards,
Andreas

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/511

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 29 Mar 2011 16:08 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 29 Mar 2011 17:40 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 29 Mar 2011 18:50 UTC**
It appears that the algorithm isn't properly limiting the upper bound of the array index. I suggest this replacement,

```
color := colorMap[1 + integer((size(colorMap,1)-1)*(max(T_min,min(T,T_max))-T_min)
                              / (T_max-T_min)), :];
```

since it requires only one call to _size_ (although I suspect the tool would optimize that away anyway). It's also easier to understand.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 30 Jun 2011 19:41 UTC**
Fixed in cc1ab9035aaf16438858ce75bce9de748708dce0 (according to the suggestion from Stefan and tested with the provided script)

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 31 Oct 2011 08:19 UTC**
